lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2551,139 +2553,174 @@ buildAndSignTransactionNew
        , GenChange s
        )
     => Tracer IO WalletLog
-    -> TimeInterpreter (Either PastHorizonException)
+    -> TimeInterpreter (ExceptT PastHorizonException IO)
     -> DBLayer IO s k
     -> NetworkLayer IO Block
     -> TransactionLayer k ktype SealedTx
+    -> Passphrase "user"
     -> WalletId
     -> ArgGenChange s
     -> AnyRecentEra
-    -> Passphrase "user"
     -> PreSelection
     -> TransactionCtx
     -> IO (Tx, TxMeta, UTCTime, SealedTx)
-buildAndSignTransactionNew
-    tr ti db netLayer txLayer walletId changeGen era pwd preSelection txCtx =
+buildAndSignTransactionNew tr ti db netLayer txLayer pwd walletId
+    argGenChange era preSelection txCtx = do
+    --
+    utxo@(_, wallet, _) <- throwOnError $
+        withExceptT (ExceptionConstructTx . ErrConstructTxNoSuchWallet) $
+            readWalletUTxOIndex @_ @_ @k db walletId
+
+    pureTimeInterpreter <- snapshot ti
+    protocolParams <- currentProtocolParameters netLayer
+
+    (tx, meta, sealed) <- throwOnError $
+        let wrapError = ExceptionWitnessTx . ErrWitnessTxWithRootKey
+         in withRootKey db walletId pwd wrapError $ \rootKey scheme ->
+                withExceptT (either ExceptionBalanceTx ExceptionConstructTx) $
+                    buildAndSignTransactionPure @k @ktype @s @n
```

Comment:
Is the idea to call updateDBMaybe around here in this PR, instead of using the `readWalletUTxOIndex`?

Reply:
Yes